### PR TITLE
fix: Search sorting

### LINKED ISSUE #1354: Latest / oldest sorting not working properly on proposals search
Reproduction steps:
* Go to https://governance.decentraland.org
* Search "0x511a22cDd2c4eE8357bB02df2578037Ffe8a4d8d"
* Check proposals

Image

### PR DESCRIPTION
**Pull request by ncomerci**
This closes #1354

### CODE REVIEW

File: src/back/routes/proposal.ts [hunk 1]
```diff
@@ -118,15 +120,15 @@ export default routes((route) => {
 
 export async function getProposals(req: WithAuth) {
   const query = req.query
-  const type = query.type && String(query.type)
-  const subtype = query.subtype && String(query.subtype)
-  const status = query.status && String(query.status)
+  const type = query.type ? (String(query.type) as ProposalType) : undefined
+  const subtype = query.subtype ? (String(query.subtype) as SubtypeOptions) : undefined
+  const status = query.status ? (String(query.status) as ProposalStatus) : undefined
```
Comment (andyesp) [inline src/back/routes/proposal.ts:125]: Can't you use `toProposalType` and `toProposalStatus` here?

Comment (andyesp) [reply src/back/routes/proposal.ts:125]: `const type = toProposalType(String(query.type)) ?? undefined`

File: src/components/Search/SortingMenu.tsx [hunk 2]
```diff
@@ -16,8 +17,12 @@ const SORT_KEY = 'order'
...
   const location = useLocation()
   const params = useMemo(() => new URLSearchParams(location.search), [location.search])
-  const order = useMemo(() => (params.get('order') === 'ASC' ? 'ASC' : 'DESC'), [params])
-  const arrowDirection = useMemo(() => (order === 'ASC' ? 'Upwards' : 'Downwards'), [order])
+  const isSearching = useMemo(() => !!params.get('search'), [params])
+  const order = useMemo(
+    () => toSearchSorting(params.get('order')) ?? (isSearching ? 'RELEVANCE' : SortingOrder.DESC),
+    [isSearching, params]
+  )
+  const arrowDirection = useMemo(() => (order === SortingOrder.ASC ? 'Upwards' : 'Downwards'), [order])
```
Comment (andyesp) [inline src/components/Search/SortingMenu.tsx:25]: Memo not necessary here (and it's possible that is also not necessary in the other variables)

File: src/hooks/useProposalsSearchParams.ts [hunk 3]
```diff
@@ -14,7 +14,7 @@ export type SearchParams = {
   search: string
   searching: boolean
   timeFrame: string
-  order: 'ASC' | 'DESC' | undefined
+  order?: SortingOrder | 'RELEVANCE'
```
Comment (1emu) [inline src/hooks/useProposalsSearchParams.ts:17]: I think we could benefit from representing the `SortingCriteria` (timeframe, relevance) as something different from the `SortingOrder` (asc, desc) 🙏

File: src/entities/Proposal/types.ts [hunk 4]
```diff
@@ -152,6 +173,17 @@ export function toPoiType(value: string | null | undefined): PoiType | null {
   return isPoiType(value) ? (value as PoiType) : null
 }
 
+export function toSearchSorting(value: string | null | undefined): 'RELEVANCE' | SortingOrder | undefined {
+  switch (value) {
+    case 'RELEVANCE':
```
Comment (andyesp) [inline src/entities/Proposal/types.ts:178]: Shouldn't Relevance be part of the sorting order?

File: src/entities/Proposal/model.ts [hunk 5]
```diff
@@ -358,13 +350,13 @@ export default class ProposalModel extends Model<ProposalAttributes> {
 
     const timeFrame = this.parseTimeframe(filter.timeFrame)
     const timeFrameKey = filter.timeFrameKey || 'created_at'
-    const orderDirection = order || 'DESC'
+    const orderDirection = !order || order === 'RELEVANCE' ? SortingOrder.DESC : order
```
Comment (andyesp) [inline src/entities/Proposal/model.ts:353]: What does `!order || order === 'RELEVANCE'` means? Could you move that to a variable? You can reuse it here and in line 359.

File: src/entities/Proposal/model.ts [hunk 6]
```diff
@@ -358,13 +350,13 @@ export default class ProposalModel extends Model<ProposalAttributes> {
 
     const timeFrame = this.parseTimeframe(filter.timeFrame)
     const timeFrameKey = filter.timeFrameKey || 'created_at'
-    const orderDirection = order || 'DESC'
+    const orderDirection = !order ? SortingOrder.DESC : order
```
Comment (andyesp) [inline src/entities/Proposal/model.ts:353]: I think it's the same to write `const orderDirection = order || SortingOrder.DESC` 🤔

File: src/entities/Proposal/utils.ts [hunk 7]
```diff
@@ -175,8 +186,32 @@ export function isProposalStatus(value: string | null | undefined): boolean {
...
+export function toProposalType<OrElse>(value: string | null | undefined, orElse: () => OrElse) {
+  return toCustomType<ProposalType, OrElse, typeof value>(value, isProposalType, orElse)
+}
+
+export function toPoiType<OrElse>(value: string | null | undefined, orElse: () => OrElse) {
+  return toCustomType<PoiType, OrElse, typeof value>(value, isPoiType, orElse)
+}
+
+export function toSortingOrder<OrElse>(value: string | null | undefined, orElse: () => OrElse) {
+  return toCustomType<SortingOrder, OrElse, typeof value>(value, isSortingOrder, orElse)
```
Comment (andyesp) [inline src/entities/Proposal/utils.ts:214]: I don't like the orElse callback. I'd change this to default value. But we can discuss this later and change it in another PR.

**Review by andyesp — APPROVED** [review]

File: src/entities/Proposal/types.ts [hunk 8]
```diff
@@ -68,6 +69,21 @@ export type ProposalAttributes<C extends Record<string, unknown> = any> = {
   textsearch: SQLStatement | string | null | undefined
 }
 
+export type FilterProposalList = {
```
Comment (1emu) [inline src/entities/Proposal/types.ts:72]: `ProposalListFilter`?

**Review by 1emu — APPROVED** [review]

**Review by 1emu — APPROVED** [review]

**Review by andyesp — APPROVED** [review]

**Pull request merged**